ZFSin/zfs/module/zfs/zfs_vnops_windows.c

```diff
@@ -4774,76 +4781,70 @@ dispatcher(
 	}
 #endif
 	validity_check = *((uint64_t *)Irp);
+	IrpSp = IoGetCurrentIrpStackLocation(Irp);
+
+	dprintf("%s: enter: major %d: minor %d: %s: type 0x%x\n", __func__, IrpSp->MajorFunction, IrpSp->MinorFunction,
+		major2str(IrpSp->MajorFunction, IrpSp->MinorFunction), Irp->Type);
 
-	FsRtlEnterFileSystem();
 
 	if (IoGetTopLevelIrp() == NULL) {
 		IoSetTopLevelIrp(Irp);
 		TopLevel = TRUE;
 	}
 
-	IrpSp = IoGetCurrentIrpStackLocation(Irp);
-
-
-	dprintf("%s: enter: major %d: minor %d: %s: type 0x%x\n", __func__, IrpSp->MajorFunction, IrpSp->MinorFunction,
-		major2str(IrpSp->MajorFunction, IrpSp->MinorFunction), Irp->Type);
-
-	Status = STATUS_NOT_IMPLEMENTED;
-
 	if (DeviceObject == ioctlDeviceObject)
 		Status = ioctlDispatcher(DeviceObject, Irp, IrpSp);
 	else {
-		mount_t *zmo = DeviceObject->DeviceExtension;
+		mount_t * zmo = DeviceObject->DeviceExtension;
 		if (zmo && zmo->type == MOUNT_TYPE_DCB)
 			Status = diskDispatcher(DeviceObject, Irp, IrpSp);
 		else if (zmo && zmo->type == MOUNT_TYPE_VCB)
 			Status = fsDispatcher(DeviceObject, Irp, IrpSp);
 		else {
-
-			extern PDRIVER_UNLOAD STOR_DriverUnload;
 			extern PDRIVER_DISPATCH STOR_MajorFunction[IRP_MJ_MAXIMUM_FUNCTION + 1];
 			if (STOR_MajorFunction[IrpSp->MajorFunction] != NULL) {
 				if (TopLevel) { IoSetTopLevelIrp(NULL); }
-				FsRtlExitFileSystem();
 				//dprintf("Relaying IRP to STORport\n");
 				return STOR_MajorFunction[IrpSp->MajorFunction](DeviceObject, Irp);
 			}
+
 			// Got a request we don't care about?
 			Status = STATUS_INVALID_DEVICE_REQUEST;
 			Irp->IoStatus.Information = 0;
 		}
 	}
 
-
-	ASSERT(validity_check == *((uint64_t *)Irp));
-
-	// IOCTL_STORAGE_GET_HOTPLUG_INFO
-	// IOCTL_DISK_CHECK_VERIFY
-	//IOCTL_STORAGE_QUERY_PROPERTY
-	Irp->IoStatus.Status = Status;
-
-	if (TopLevel) { IoSetTopLevelIrp(NULL); }
-	FsRtlExitFileSystem();
+	if (TopLevel) {
+		IoSetTopLevelIrp(NULL);
```

Comment:
I am wondering if the IoSetTopLevelIrp() / IoGetTopLevelIrp() calls should only be used when under the FsRtlEnter/Exit critical section because those 2 calls are also related to filesystem only and should then nly be used when treating FS calls (I guess).  In Jorgen's code we can see that he is using the IoSetTopLevelIrp() / IoGetTopLevelIrp() calls in between taking the critical section so we should double check.  can you research that on the web to have a definite answer and eventually move the calls whenever we take the critical section?